File: java/L900/LeetCode0993.java
package L900;

import java.util.Deque;
import java.util.LinkedList;

import util.TreeNode;

public class LeetCode0993 {
    public boolean isCousins(TreeNode root, int x, int y) {
        if (root == null) {
            return false;
        }
        Deque<TreeNode> qu = new LinkedList<>();
        qu.offer(root);
        while (!qu.isEmpty()) {
            int sz = qu.size();
            int[] flags = new int[]{-1, -1};
            for (int i = 0;i < sz;++i) {
                TreeNode node = qu.poll();
                if (node.left != null) {
                    if (node.left.val == x) {
                        if (flags[1] != -1 && flags[1] != i) {
                            return true;
                        }
                        if (flags[0] == -1) {
                            flags[0] = i;
                        }
                    } else if (node.left.val == y) {
                        if (flags[0] != -1 && flags[0] != i) {
                            return true;
                        }
                        if (flags[1] == -1) {
                            flags[1] = i;
                        }
                    }
                    qu.offer(node.left);
                }
                if (node.right != null) {
                    if (node.right.val == x) {
                        if (flags[1] != -1 && flags[1] != i) {
                            return true;
                        }
                        if (flags[0] == -1) {
                            flags[0] = i;
                        }
                    } else if (node.right.val == y) {
                        if (flags[0] != -1 && flags[0] != i) {
                            return true;
                        }
                        if (flags[1] == -1) {
                            flags[1] = i;
                        }
                    }
                    qu.offer(node.right);
                }
            }
        }
        return false;
    }
}

File: cpp/L900/LeetCode0993.cc
#include "../heads/tree_node.h"
#include <queue>

class Solution {
    bool sameParent(TreeNode *node, int x, int y) {
        return node->left != nullptr && (node->left->val == x || node->left->val == y)
         && node->right != nullptr && (node->right->val == x || node->right->val == y);
    }
public:
    bool isCousins(TreeNode* root, int x, int y) {
        if (root == nullptr) {
            return false;
        }
        std::queue<TreeNode*> q;
        q.push(root);
        int count{0};
        while (!q.empty()) {
            int qz = q.size();
            for (int i = 0;i < qz;++i) {
                TreeNode *node = q.front();
                q.pop();
                if (sameParent(node, x, y)) {
                    return false;
                }
                if (node->left != nullptr) {
                    if (node->left->val == x || node->left->val == y) {
                        ++count;
                    }
                    q.push(node->left);
                }
                if (node->right != nullptr) {
                    if (node->right->val == x || node->right->val == y) {
                        ++count;
                    }
                    q.push(node->right);
                }
            }
            if (count == 2) {
                return true;
            }
            count = 0;
        }
        return false;
    }
};